{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "82300ca4",
   "metadata": {},
   "source": [
    "## A survey of iodine transporters, for identification of candidates for recombination into a microbe.\n",
    "\n",
    "Procedure:\n",
    "- scrape human, mouse [Na-I Sympoters](https://en.wikipedia.org/wiki/Sodium/iodide_cotransporter)\n",
    "- verify that they have conserved sequences\n",
    "- scrape for other human, mouse [solute carrier membrane transport proteins](https://en.wikipedia.org/wiki/Solute_carrier_family). Do it to get a sense of what the active domains might be that make the Na-I symporter specific for iodine.\n",
    "- loop:\n",
    "  - identify peptide sequences of interest.\n",
    "  - search databases for other matching domains.\n",
    "  - if results are satisfactory and the crucial domains are clear, exit loop.\n",
    "  - else, expand the search to other species up the phylogenetic tree (moving in the direction of microbes and kelp) and look at other proteins across life that are specific to iodine, by doing literature research.\n",
    "  - identify candidate species proteins, and find sequencing data.\n",
    "- end with a protein of interest that is a likely candidate for being a pump that's viable for "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02ef1d96",
   "metadata": {},
   "source": [
    "General links:\n",
    "- [Capnocytophaga canimorsus](https://en.wikipedia.org/wiki/Capnocytophaga_canimorsus)\n",
    "- [Na-I Symporter](https://en.wikipedia.org/wiki/Sodium/iodide_cotransporter)\n",
    "- [Solute Carrier Family](https://en.wikipedia.org/wiki/Solute_carrier_family)\n",
    "- [Iodine in biology](https://en.wikipedia.org/wiki/Iodine_in_biology)\n",
    "- [Iodine](https://en.wikipedia.org/wiki/Iodine#Biological_role)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "55f40b47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. Executable: /home/morgan/projects/bioinfo-projects/comparative-sequence-analysis/.venv/bin/python\n",
      "2. Version: 3.11.2 (main, Apr 28 2025, 14:11:48) [GCC 12.2.0]\n",
      "3. Env Path: /home/morgan/projects/bioinfo-projects/comparative-sequence-analysis/.venv\n"
     ]
    }
   ],
   "source": [
    "# Run this code inside a Jupyter cell. If all three lines point to your .venv folder in WSL, your setup is perfect:\n",
    "# In this project (iodine transporter survey), the venv should be found in ../comparative-sequence-analysis, \n",
    "# should we need to get it manually.\n",
    "\n",
    "import sys, os\n",
    "print(f\"1. Executable: {sys.executable}\") \n",
    "print(f\"2. Version: {sys.version}\")\n",
    "print(f\"3. Env Path: {os.getenv('VIRTUAL_ENV')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1daafe6",
   "metadata": {},
   "source": [
    "### SLC5A5 / NIS / TDH1 - solute carrier family 5 member 5 - gene id 6528\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/datasets/gene/6528/\n",
    "- RefSeq summary:     This gene encodes a member of the sodium glucose cotransporter family. The encoded protein is responsible for the uptake of iodine in tissues such as the thyroid and lactating breast tissue. The iodine taken up by the thyroid is incorporated into the metabolic regulators triiodothyronine (T3) and tetraiodothyronine (T4). Mutations in this gene are associated with thyroid dyshormonogenesis 1.[provided by RefSeq, Sep 2009]\n",
    "- 8 proteins, 8 transcipts, downloaded as ncbi_dataset(1).zip https://www.ncbi.nlm.nih.gov/datasets/gene/6528/#transcripts-and-proteins\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/gene?cmd=retrieve&dopt=default&rn=1&list_uids=6528\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/KAI4041341.1 (partial) sequence.gp\n",
    "https://www.ncbi.nlm.nih.gov/protein/KAI2589675.1 (partial?) sequence(1).gp\n",
    "\n",
    "https://www.uniprot.org/uniprot/Q92911"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841046e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='NP_000444.1', name='NP_000444.1', description='NP_000444.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='NP_001427636.1', name='NP_001427636.1', description='NP_001427636.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='XP_011526494.1', name='XP_011526494.1', description='XP_011526494.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='XP_011526495.1', name='XP_011526495.1', description='XP_011526495.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRFSRAVRLCGTLQYIVATMLYTGIVIYAPALILNQVTGLDIWASLLSTGIICT...TNL'), id='XP_011526496.1', name='XP_011526496.1', description='XP_011526496.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X4]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='XP_054177777.1', name='XP_054177777.1', description='XP_054177777.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='XP_054177778.1', name='XP_054177778.1', description='XP_054177778.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRFSRAVRLCGTLQYIVATMLYTGIVIYAPALILNQVTGLDIWASLLSTGIICT...TNL'), id='XP_054177780.1', name='XP_054177780.1', description='XP_054177780.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X4]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='KAI4041341.1', name='KAI4041341.1', description='KAI4041341.1 solute carrier family 5 member 5 [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='KAI2589675.1', name='KAI2589675.1', description='KAI2589675.1 solute carrier family 5 member 5 [Homo sapiens]', dbxrefs=[])]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequences =[]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/ncbi_dataset/data/protein.faa', \"fasta\"):\n",
    "    sequences += [record]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/sequence.fasta', 'fasta'):\n",
    "    sequences += [record]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/sequence(1).fasta', 'fasta'):\n",
    "    sequences += [record]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "3a8832da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[643., 554., 632., 543., 510., 632., 543., 510., 643., 643.],\n",
       "       [ inf, 554., 543., 543., 510., 543., 543., 510., 554., 554.],\n",
       "       [ inf,  inf, 654., 565., 532., 654., 565., 532., 632., 632.],\n",
       "       [ inf,  inf,  inf, 565., 532., 565., 565., 532., 543., 543.],\n",
       "       [ inf,  inf,  inf,  inf, 532., 532., 532., 532., 510., 510.],\n",
       "       [ inf,  inf,  inf,  inf,  inf, 654., 565., 532., 632., 632.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf, 565., 532., 543., 543.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf, 532., 510., 510.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf, 643., 643.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf]])"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from Bio import Align\n",
    "import numpy as np\n",
    "arr = np.ones([len(sequences),len(sequences)])*np.inf\n",
    "\n",
    "aligner = Align.PairwiseAligner(mode='local') # local is better than global or fogsaa in this task of long sequences.\n",
    "all_aligns = {}\n",
    "for si1 in range(len(sequences)-1):\n",
    "  for si2 in range(si1, len(sequences)):\n",
    "    alignments = aligner.align(sequences[si1], sequences[si2])\n",
    "    # print('='*80)\n",
    "    # print(si1,si2)\n",
    "    # print(alignments)\n",
    "    # print(alignments[0])\n",
    "    all_aligns[(si1,si2)] = alignments[0]\n",
    "    arr[si1,si2] = alignments[0].score\n",
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "a1a5cf0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[133.  44. 122.  33.   0. 122.  33.   0. 133. 133.]\n",
      " [ inf  44.  33.  33.   0.  33.  33.   0.  44.  44.]\n",
      " [ inf  inf 144.  55.  22. 144.  55.  22. 122. 122.]\n",
      " [ inf  inf  inf  55.  22.  55.  55.  22.  33.  33.]\n",
      " [ inf  inf  inf  inf  22.  22.  22.  22.   0.   0.]\n",
      " [ inf  inf  inf  inf  inf 144.  55.  22. 122. 122.]\n",
      " [ inf  inf  inf  inf  inf  inf  55.  22.  33.  33.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  22.   0.   0.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  inf 133. 133.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  inf  inf  inf]]\n"
     ]
    }
   ],
   "source": [
    "print(arr-np.min(arr))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8605d72e",
   "metadata": {},
   "source": [
    "By this analysis, it seems small chunks are excised, which might even be a facet of sequencing. There aren't small mutations that can show differences in function -- which makes sense among individuals in the same species. Todo, look at mouse and other species data."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv (3.11.2)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
