{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "82300ca4",
   "metadata": {},
   "source": [
    "## A survey of iodine transporters, for identification of candidates for recombination into a microbe.\n",
    "\n",
    "Procedure:\n",
    "- scrape human, mouse [Na-I Sympoters](https://en.wikipedia.org/wiki/Sodium/iodide_cotransporter)\n",
    "- verify that they have conserved sequences\n",
    "- scrape for other human, mouse [solute carrier membrane transport proteins](https://en.wikipedia.org/wiki/Solute_carrier_family). Do it to get a sense of what the active domains might be that make the Na-I symporter specific for iodine.\n",
    "- loop:\n",
    "  - identify peptide sequences of interest.\n",
    "  - search databases for other matching domains.\n",
    "  - if results are satisfactory and the crucial domains are clear, exit loop.\n",
    "  - else, expand the search to other species up the phylogenetic tree (moving in the direction of microbes and kelp) and look at other proteins across life that are specific to iodine, by doing literature research.\n",
    "  - identify candidate species proteins, and find sequencing data.\n",
    "- end with a protein of interest that is a likely candidate for being a pump that's viable for "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02ef1d96",
   "metadata": {},
   "source": [
    "General links:\n",
    "- [Capnocytophaga canimorsus](https://en.wikipedia.org/wiki/Capnocytophaga_canimorsus)\n",
    "- [Na-I Symporter](https://en.wikipedia.org/wiki/Sodium/iodide_cotransporter)\n",
    "- [Solute Carrier Family](https://en.wikipedia.org/wiki/Solute_carrier_family)\n",
    "- [Iodine in biology](https://en.wikipedia.org/wiki/Iodine_in_biology)\n",
    "- [Iodine](https://en.wikipedia.org/wiki/Iodine#Biological_role)\n",
    "- Nitrate from [Nitratine](https://en.wikipedia.org/wiki/Nitratine) \"Nitratine ... has been known since 1845 from mineral deposits in the Confidence Hills, Southern Death Valley, California and the Atacama Desert, Chile. It is still used in organic farming (where Haber–Bosch ammonia is forbidden) in the US, but prohibited in international organic agriculture.[9]\"\n",
    "- Fertilizer constituents [(src)](https://en.wikipedia.org/wiki/Fertilizer): Fertilizers typically provide, in varying proportions:[35] Three main macronutrients (NPK):\n",
    "        Nitrogen (N): leaf growth and stems;[36]\n",
    "        Phosphorus (P): development of roots, flowers, seeds and fruit;[37]\n",
    "        Potassium (K): strong stem growth, movement of water in plants, promotion of flowering and fruiting;[38]\n",
    "    Three secondary macronutrients: calcium (Ca), magnesium (Mg), and sulfur (S);[39]\n",
    "    Micronutrients: copper (Cu), iron (Fe), manganese (Mn), molybdenum (Mo), zinc (Zn), and boron (B).[40] Of occasional significance are silicon (Si), cobalt (Co), and vanadium (V)\n",
    "- [Eutrophication](https://en.wikipedia.org/wiki/Eutrophication)\n",
    "- Productive Species: Laminaria digitata , Laminaria hyperborea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "55f40b47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. Executable: /home/morgan/projects/bioinfo-projects/comparative-sequence-analysis/.venv/bin/python\n",
      "2. Version: 3.11.2 (main, Apr 28 2025, 14:11:48) [GCC 12.2.0]\n",
      "3. Env Path: /home/morgan/projects/bioinfo-projects/comparative-sequence-analysis/.venv\n"
     ]
    }
   ],
   "source": [
    "# Run this code inside a Jupyter cell. If all three lines point to your .venv folder in WSL, your setup is perfect:\n",
    "# In this project (iodine transporter survey), the venv should be found in ../comparative-sequence-analysis, \n",
    "# should we need to get it manually.\n",
    "\n",
    "import sys, os\n",
    "print(f\"1. Executable: {sys.executable}\") \n",
    "print(f\"2. Version: {sys.version}\")\n",
    "print(f\"3. Env Path: {os.getenv('VIRTUAL_ENV')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1daafe6",
   "metadata": {},
   "source": [
    "### SLC5A5 / NIS / TDH1 - solute carrier family 5 member 5 - gene id 6528\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/datasets/gene/6528/\n",
    "- RefSeq summary:     This gene encodes a member of the sodium glucose cotransporter family. The encoded protein is responsible for the uptake of iodine in tissues such as the thyroid and lactating breast tissue. The iodine taken up by the thyroid is incorporated into the metabolic regulators triiodothyronine (T3) and tetraiodothyronine (T4). Mutations in this gene are associated with thyroid dyshormonogenesis 1.[provided by RefSeq, Sep 2009]\n",
    "- 8 proteins, 8 transcipts, downloaded as ncbi_dataset(1).zip https://www.ncbi.nlm.nih.gov/datasets/gene/6528/#transcripts-and-proteins\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/gene?cmd=retrieve&dopt=default&rn=1&list_uids=6528\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/KAI4041341.1 (partial) sequence.gp\n",
    "https://www.ncbi.nlm.nih.gov/protein/KAI2589675.1 (partial?) sequence(1).gp\n",
    "\n",
    "https://www.uniprot.org/uniprot/Q92911\n",
    "[this chart here](https://www.uniprot.org/uniprotkb/Q92911/variant-viewer) is an excellent way to identify regions of interest, especially after matching it to conserved regions, and/or seeing if this region appears in other distant species. [this](https://www.uniprot.org/uniprotkb?query=iodine+NOT+%28taxonomy_id%3A33208%29) and [this](https://www.uniprot.org/uniprotkb?query=iodine+AND+transporter+NOT+%28taxonomy_id%3A33208%29) searches for not metazoa && iodine (&& transporter).\n",
    "\n",
    "About medical: Preedy V (2009). Comprehensive Handbook of Iodine Nutritional, Biochemical, Pathological and Therapeutic Aspects. Burlington: Elsevier. p. 616. ISBN 978-0-08-092086-3. [link](https://books.google.com/books?id=7v7g5XoCQQwC&pg=PA616). [src wiki hypothyroidism](https://en.wikipedia.org/wiki/Hypothyroidism#Prevention)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841046e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='NP_000444.1', name='NP_000444.1', description='NP_000444.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='NP_001427636.1', name='NP_001427636.1', description='NP_001427636.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='XP_011526494.1', name='XP_011526494.1', description='XP_011526494.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='XP_011526495.1', name='XP_011526495.1', description='XP_011526495.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRFSRAVRLCGTLQYIVATMLYTGIVIYAPALILNQVTGLDIWASLLSTGIICT...TNL'), id='XP_011526496.1', name='XP_011526496.1', description='XP_011526496.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X4]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='XP_054177777.1', name='XP_054177777.1', description='XP_054177777.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X1]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MCLGQLLNSVLTALLFMPVFYRLGLTSTYEYLEMRFSRAVRLCGTLQYIVATML...TNL'), id='XP_054177778.1', name='XP_054177778.1', description='XP_054177778.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X2]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MRFSRAVRLCGTLQYIVATMLYTGIVIYAPALILNQVTGLDIWASLLSTGIICT...TNL'), id='XP_054177780.1', name='XP_054177780.1', description='XP_054177780.1 SLC5A5 [organism=Homo sapiens] [GeneID=6528] [isoform=X4]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='KAI4041341.1', name='KAI4041341.1', description='KAI4041341.1 solute carrier family 5 member 5 [Homo sapiens]', dbxrefs=[]),\n",
       " SeqRecord(seq=Seq('MEAVETGERPTFGAWDYGVFALMLLVSTGIGLWVGLARGGQRSAEDFFTGGRRL...TNL'), id='KAI2589675.1', name='KAI2589675.1', description='KAI2589675.1 solute carrier family 5 member 5 [Homo sapiens]', dbxrefs=[])]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "sequences =[]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/ncbi_dataset/data/protein.faa', \"fasta\"):\n",
    "    sequences += [record]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/sequence.fasta', 'fasta'):\n",
    "    sequences += [record]\n",
    "for record in SeqIO.parse('data/SLC5A5-h-sapiens/sequence(1).fasta', 'fasta'):\n",
    "    sequences += [record]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "3a8832da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[643., 554., 632., 543., 510., 632., 543., 510., 643., 643.],\n",
       "       [ inf, 554., 543., 543., 510., 543., 543., 510., 554., 554.],\n",
       "       [ inf,  inf, 654., 565., 532., 654., 565., 532., 632., 632.],\n",
       "       [ inf,  inf,  inf, 565., 532., 565., 565., 532., 543., 543.],\n",
       "       [ inf,  inf,  inf,  inf, 532., 532., 532., 532., 510., 510.],\n",
       "       [ inf,  inf,  inf,  inf,  inf, 654., 565., 532., 632., 632.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf, 565., 532., 543., 543.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf, 532., 510., 510.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf, 643., 643.],\n",
       "       [ inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf,  inf]])"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from Bio import Align\n",
    "import numpy as np\n",
    "arr = np.ones([len(sequences),len(sequences)])*np.inf\n",
    "\n",
    "aligner = Align.PairwiseAligner(mode='local') # local is better than global or fogsaa in this task of long sequences.\n",
    "all_aligns = {}\n",
    "for si1 in range(len(sequences)-1):\n",
    "  for si2 in range(si1, len(sequences)):\n",
    "    alignments = aligner.align(sequences[si1], sequences[si2])\n",
    "    # print('='*80)\n",
    "    # print(si1,si2)\n",
    "    # print(alignments)\n",
    "    # print(alignments[0])\n",
    "    all_aligns[(si1,si2)] = alignments[0]\n",
    "    arr[si1,si2] = alignments[0].score\n",
    "arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "a1a5cf0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[133.  44. 122.  33.   0. 122.  33.   0. 133. 133.]\n",
      " [ inf  44.  33.  33.   0.  33.  33.   0.  44.  44.]\n",
      " [ inf  inf 144.  55.  22. 144.  55.  22. 122. 122.]\n",
      " [ inf  inf  inf  55.  22.  55.  55.  22.  33.  33.]\n",
      " [ inf  inf  inf  inf  22.  22.  22.  22.   0.   0.]\n",
      " [ inf  inf  inf  inf  inf 144.  55.  22. 122. 122.]\n",
      " [ inf  inf  inf  inf  inf  inf  55.  22.  33.  33.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  22.   0.   0.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  inf 133. 133.]\n",
      " [ inf  inf  inf  inf  inf  inf  inf  inf  inf  inf]]\n"
     ]
    }
   ],
   "source": [
    "print(arr-np.min(arr))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8605d72e",
   "metadata": {},
   "source": [
    "By this analysis, it seems small chunks are excised, which might even be a facet of sequencing. There aren't small mutations that can show differences in function -- which makes sense among individuals in the same species. Todo, look at mouse and other species data."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "802fae3c",
   "metadata": {},
   "source": [
    "##  Reviewing possible Iodine transporters\n",
    "Surveying the basic options for iodine transport seems like a good first step.\n",
    "This list begins to compile the ways organisms transport iodine across membranes. \n",
    "For now, I'll exclude iodine metabolism processing, but that'll likely be important:\n",
    "- A microbe that is engineered to intake a vast concentration of iodine will likely be poisoned by it\n",
    "- Therefore, some means of sequestering the iodine is expected to be necessary.\n",
    "- Two options that come to mind are:\n",
    "  1) attaching it to free tyrosine and stacking that tyrosine either as a oligotyrosine peptide or free-floating iodo-tyrosines\n",
    "  2) put the iodide ions directly into an inclusion body\n",
    "  The first method suggests engineering metabolic enzymes. The second suggests vacuole transporters or working with organisms that isolate salt into deposits. These aren't the main focus, but I may as well include them at the bottom of the list. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "298f7f0e",
   "metadata": {},
   "source": [
    "Main two categories of ways to get iodine into protists:\n",
    "### Iodine transporters \n",
    "1) Na-I symporter, human mammal. SLC5A5\n",
    "1) Na-Br symporter, human\n",
    "\n",
    "### (vanadium-dependent halo)peroxidases.\n",
    "1) vanadium-dependent iodine peroxidase (vIPO), laminaria kelp.\n",
    "2) (vHPO) vanadium-dependent haloperoxidase, diatom\n",
    "3) vHPO, murex snail, sea urchin, horseradish, etc: see web-browsing-infodumps.md [tag514432] for more species\n",
    "\n",
    "### Other proteins and mechanisms of interest:\n",
    "1) MCT (monocarboxylate transporter) SLC16A2, human mammal. (This transports thyroid hormones, compounds built on iodinated tyrosine; thyroid hormones cannot cross plasma membranes on their own). (from [wiki](https://en.wikipedia.org/wiki/Monocarboxylate_transporter_8), see also zebrafish and frogs  Xenopus laevis and Xenopus tropicalis.)\n",
    "2) iodine transferases\n",
    "3) Pendrin, human mammal.\n",
    "4) transporter of thyroxine-binding globulin, human mammal\n",
    "5) transporter of transthyretin, thyroxine-binding albumin, human mammal. (Lower binding strength than thyroxine-binding globulin; [src](https://en.wikipedia.org/wiki/Transthyretin)\n",
    "6) Thyroglobulin (Tg), storage protein of iodine and inactive thyroid hormone and catalyzes their bonding\n",
    "   - thyroglobulin analogues in brown algae, protists?\n",
    "7) Deiodinase, selenium-dependent, human\n",
    "8) proteins in the endostyles of lampreys, amphioxi, ascidia, tunicate, Lancelet. see <feb-10-workcomp>, & maybe [tag514432]\n",
    "9) human peroxiredoxin genes: PRDX1, ... PRDX6.   \n",
    "10) ai google for \"halo transferase\":  see below [tagHaloTag]\n",
    "11) thyroid peroxidase. [src](https://en.wikipedia.org/wiki/Thyroid_hormones):\n",
    "    - This iodide enters the follicular lumen from the cytoplasm by the transporter pendrin, in a purportedly passive manner.\n",
    "    - In the colloid, iodide (I−) is oxidized to iodine (I0) by an enzyme called **thyroid peroxidase**.\n",
    "    - Iodine (I0) is very reactive and iodinates the thyroglobulin at tyrosyl residues in its protein chain (in total containing approximately 120 tyrosyl residues). \n",
    "12) https://en.wikipedia.org/wiki/Recombinant_human_parathyroid_hormone\n",
    "13) https://en.wikipedia.org/wiki/Thyroxine_5-deiodinase, esp @ releasing the iodine in response to stress\n",
    "14) [Thyroid peroxidase](https://en.wikipedia.org/wiki/Thyroid_peroxidase), human: I- -> I0 for addition to tyrosine, encoded by TPO gene. aka thyroperoxidase (TPO), thyroid specific peroxidase or iodide peroxidase.\n",
    "15) [Megalin](https://en.wikipedia.org/wiki/Megalin) aka Low density lipoprotein receptor-related protein 2 also known as LRP-2 , LRP-2 gene. \"LRP2 is expressed in epithelial cells of the thyroid (thyrocytes), where it can serve as a receptor for the protein thyroglobulin (Tg).\"  [src](https://en.wikipedia.org/wiki/Thyroid_hormones): \"Iodinated thyroglobulin binds megalin for endocytosis back into the cell.\"\n",
    "16) meta: filter proteins by their [protein targeting](https://en.wikipedia.org/wiki/Protein_targeting) sequences, to see where they go. ie, in algae, look at proteins tend to find their way to near the membrane. \n",
    "17) meta: find a thyroglobulin alternative by seeing what if anything in algae has lots of tyr sequences and no other apparent function.\n",
    "\n",
    "\n",
    "\n",
    "### Organisms\n",
    "- bacillus subtilis\n",
    "- homo sapiens\n",
    "- laminara digitalis\n",
    "\n",
    "### Bioinformatics analogue search fronts\n",
    "- peroxidase (vanadium and otherwise)\n",
    "- halotransporters\n",
    "- thyroglobulin \n",
    "- dehalogenase "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2d2542c",
   "metadata": {},
   "source": [
    "-- ----\n",
    "\n",
    "procedure 1: vHPO + cytoplasmic thyroglobulin\n",
    "1) use vanadium-dependent haloperoxidase to deionize iodine so it passes into the organism, as atom or molecule\n",
    "2) this reactive I0 species will tend to react with whatever's near, nonspecifically. Provide it thyroglobulin, or other tyrosine-rich peptide source, that can soak up reactive iodine, ideally near outer membrane.\n",
    "- more native, but less controlled\n",
    "\n",
    "procedure 2: Na/I symporter + TPO + thyroglobulin\n",
    "1) use human Na/I symporter SLC5A5 to transport iodide ions\n",
    "2) oxidize the ions into I0 at the right location with TPO, possibly\n",
    "3) let them react to tyrosine with thyroglobulin\n",
    "- less native, but more directed\n",
    "\n",
    "extras:\n",
    "1) eliminate any destination metabolic processes that utilize iodinated tyrosyls.\n",
    "2) if need be, provide an internal vesicle to pump the I0 into via pendrin or Na/I symporter.\n",
    "3) for extraction, consider looking into what laminara/etc uses to release the I in response to stress. perhaps the right antigens can cause it. \n",
    "\n",
    "-- ----\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db034e6e",
   "metadata": {},
   "source": [
    "[tagHaloTag]\n",
    "\n",
    "A halotransferase, often referred to in the context of HaloTag technology, is a modified haloalkane dehalogenase enzyme designed to covalently and irreversibly bind to specific chloroalkane-linked ligands. It serves as a versatile, self-labeling protein tag used for molecular imaging, protein purification, and immobilization, primarily enabling cellular, subcellular, and, in some cases, in-soil visualization of proteins. [1, 2, 3]  \n",
    "Key Aspects of HaloTag/Halotransferase Technology: \n",
    "\n",
    "• Mechanism: The engineered enzyme (HaloTag protein) reacts rapidly under physiological conditions with synthetic ligands, forming a covalent bond with a chloroalkane linker. \n",
    "• Applications: It allows for tagging proteins with various fluorescent dyes, affinity handles (like biotin), or solid supports. \n",
    "• Utility: Used extensively in, for example, this study (https://pubs.acs.org/doi/10.1021/cb800025k) for imaging NF-κB cellular translocation and studying protein-protein or protein-DNA complexes. \n",
    "• Versatility: A single Genetic fusion allows for different functional ligands to be applied, note this patent (https://patents.google.com/patent/EP3891226A2/en). \n",
    "• Distinction: It is distinct from metabolic enzymes that break down anesthetics, such as those discussed in this NIH study (https://pubmed.ncbi.nlm.nih.gov/3364763/). [1, 2, 4, 5]  \n",
    "\n",
    "It should not be confused with HALO-tag which is often used interchangeably in molecular biology. \n",
    "\n",
    "AI responses may include mistakes.\n",
    "\n",
    "- [1] https://pubs.acs.org/doi/10.1021/cb800025k\n",
    "- [2] https://pmc.ncbi.nlm.nih.gov/articles/PMC3480824/\n",
    "- [3] https://pubs.acs.org/doi/10.1021/acs.est.6b01415\n",
    "- [4] https://pubmed.ncbi.nlm.nih.gov/3364763/\n",
    "- [5] https://patents.google.com/patent/EP3891226A2/en\n",
    "\n",
    "see also:\n",
    "\n",
    " https://pubs.acs.org/doi/10.1021/acs.est.6b01415\n",
    " Volatile Gas Production by Methyl Halide Transferase: An In Situ Reporter Of Microbial Gene Expression In Soil\n",
    " 2016,  42 cites,    Hsiao-Ying Cheng, Caroline A. Masiello, George N. Bennett, Jonathan J. Silberg"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv (3.11.2)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
